Update schemas for version 2.3

goneall opened the pull request:
See the list of changes in [this comment for issue 691](https://github.com/spdx/spdx-spec/issues/691#issuecomment-1154198748).

Note that not all of the items listed in the comment needed updates.

This PR also copies over fixes from PR #547 

Signed-off-by: Gary O'Neall <gary@sourceauditor.com>

goneall commented:
Going back through my checklist I found some text for 2.2.1 that was not included in this update.

Changing to dreaft.

goneall commented:
This now include updates to fix #459

goneall commented:
Fixes #606

davaya commented:
JSON schema $id should identify the schema (as a purl identifies a package).

From: https://json-schema.org/understanding-json-schema/structuring.html:
> Schema documents are identified by non-relative URIs.

Line 2 of schemas/spdx-schema.json should be changed from a constant to a unique namespace identifier, something like:

`"$id" : "http://spdx.org/rdf/terms"`  
`"$id" : "http://spdx.org/spdx/v2.3",`

goneall commented:
@fviernau - Can you take a look and see if we missed anything?

zvr commented on ontology/spdx-ontology.owl.xml:
```diff
@@ -9,14 +9,15 @@
-Known issues:</rdfs:comment>
-        <rdfs:label>SPDX 2.2</rdfs:label>
-        <owl:versionInfo rdf:datatype="http://www.w3.org/2001/XMLSchema#decimal">2.2</owl:versionInfo>
+        <owl:versionIRI rdf:resource="http://spdx.org/rdf/terms/2.3"/>
```
I'd use the simpler `http://spdx.org/rdf/2.3`

goneall replied:
> I'd use the simpler http://spdx.org/rdf/2.3

Previous versions have always used the `rdf/terms` as part of the URI - so to keep it consistent, I would like to leave this as is.

zvr commented on ontology/spdx-ontology.owl.xml:
```diff
@@ -9,14 +9,15 @@
+Known issues:
+- rdfs:comment and rdfs:seeAlso are used within the SPDX classes and causes a redefinition of these standard terms</rdfs:comment>
+        <rdfs:label>SPDX 2.3</rdfs:label>
+        <owl:versionInfo rdf:datatype="http://www.w3.org/2001/XMLSchema#decimal">2.3</owl:versionInfo>
```
Why is the version a decimal? I'd suggest the datatype is deleted, by default it's a string.

goneall replied:
@zvr I agree - however, whenever I try to change this in Protege tool, it keeps changing it back to a decimal type.  I looked up the W3C OWL spec and there is no requirement for this to be a decimal datatype.

Let me know if you find a way to fix this (other than manually patching it after editing in Protege).

zvr commented on ontology/spdx-ontology.owl.xml:
```diff
@@ -307,7 +312,9 @@ Known issues:</rdfs:comment>
             </owl:Class>
         </rdfs:range>
-        <rdfs:comment xml:lang="en">The licensing that the preparer of this SPDX document has concluded, based on the evidence, actually applies to the package.</rdfs:comment>
+        <rdfs:comment xml:lang="en">The licensing that the preparer of this SPDX document has concluded, based on the evidence, actually applies to the package.
```
`licenseConcluded` is not only for packages, but also for files and snippets

zvr commented on ontology/spdx-ontology.owl.xml:
```diff
@@ -307,7 +312,9 @@ Known issues:</rdfs:comment>
-        <rdfs:comment xml:lang="en">The licensing that the preparer of this SPDX document has concluded, based on the evidence, actually applies to the package.</rdfs:comment>
+        <rdfs:comment xml:lang="en">The licensing that the preparer of this SPDX document has concluded, based on the evidence, actually applies to the package.
+
+If the Concluded License field is not present for an SPDX Item, it implies an equivalent meaning to NOASSERTION.</rdfs:comment>
```
Comments refer to the RDF terms, so use `licenseConcluded` instead of `Concluded License`

zvr commented on ontology/spdx-ontology.owl.xml:
```diff
@@ -354,7 +361,9 @@ Known issues:</rdfs:comment>
-        <rdfs:comment xml:lang="en">The licensing information that was discovered directly within the package. There will be an instance of this property for each distinct value of alllicenseInfoInFile properties of all files contained in the package.</rdfs:comment>
+        <rdfs:comment xml:lang="en">The licensing information that was discovered directly within the package. There will be an instance of this property for each distinct value of alllicenseInfoInFile properties of all files contained in the package.
+
+If the All Licenses Information from Files field is not present for a package and FilesAnalyzed property for that same pacakge is true or omitted, it implies an equivalent meaning to NOASSERTION.</rdfs:comment>
```
Comments refer to the RDF terms, so use `licenseInfoFromFiles` instead of `All Licenses Information from Files` (`FilesAnalyzed` is correct)

zvr commented on ontology/spdx-ontology.owl.xml:
```diff
@@ -380,8 +389,10 @@ Known issues:</rdfs:comment>
-        <ns:term_status>deprecated</ns:term_status>
+        <rdfs:comment xml:lang="en">Licensing information that was discovered directly in the subject file. This is also considered a declared license for the file.
+
+If the License Information in File field is not present for a file, it implies an equivalent meaning to NOASSERTION.</rdfs:comment>
```
Comments refer to the RDF terms, so use `licenseInfoInFile` instead of `License Information in File`

zvr commented on ontology/spdx-ontology.owl.xml:
```diff
@@ -406,7 +417,9 @@ Known issues:</rdfs:comment>
-        <rdfs:comment xml:lang="en">Licensing information that was discovered directly in the subject snippet. This is also considered a declared license for the snippet.</rdfs:comment>
+        <rdfs:comment xml:lang="en">Licensing information that was discovered directly in the subject snippet. This is also considered a declared license for the snippet.
+
+If the License Information in Snippet field is not present for a snippet, it implies an equivalent meaning to NOASSERTION.</rdfs:comment>
```
Comments refer to the RDF terms, so use `licenseInfoInSnippet` instead of `License Information in Snippet`

zvr commented on ontology/spdx-ontology.owl.xml:
```diff
@@ -786,7 +824,9 @@ Known issues:</rdfs:comment>
         <rdfs:range rdf:resource="http://www.w3.org/2000/01/rdf-schema#Literal"/>
         <rdfs:range rdf:resource="http://www.w3.org/2001/XMLSchema#string"/>
-        <rdfs:comment xml:lang="en">The text of copyright declarations recited in the Package or File.</rdfs:comment>
+        <rdfs:comment xml:lang="en">The text of copyright declarations recited in the Package, File or snippet.
```
For consistency, use `package, file` instead of `Package, File`. There are also more instances of incorrect capitalization.

goneall replied:
These should just be copy/paste from the spec.  I'll fix this one, but I probably won't have time to review the entire set of comments for capitalization.

zvr commented on ontology/spdx-ontology.owl.xml:
```diff
@@ -786,7 +824,9 @@ Known issues:</rdfs:comment>
-        <rdfs:comment xml:lang="en">The text of copyright declarations recited in the Package or File.</rdfs:comment>
+        <rdfs:comment xml:lang="en">The text of copyright declarations recited in the Package, File or snippet.
+
+If the License Information in File field is not present, it implies an equivalent meaning to NOASSERTION.</rdfs:comment>
```
Definitely use `copyrightText` instead of `License Information in File`

zvr reviewed (changes requested):
I only looked and commented on a single ontology file, the XML one. I assume all are somehow automatically cross-generated.

It appears that on GitHub I can only comment on changed lines, so here's a couple of comments on non-changed lines:

line 3, default and base namespaces: 
Shall we move this to `http://spdx.org/rdf/2.3/terms`? At some point we should, definitely for 3.0.

line 1124, `licenseId`:
According to Annex D (SPDX license expressions), we actually allow `["DocumentRef-"(idstring)":"]"LicenseRef-"(idstring)`. The description should be amended to reflect this.

rjb4standards commented on schemas/spdx-schema.json:
```diff
@@ -380,6 +385,14 @@
             "description" : "The base name of the package file name. For example, zlib-1.2.5.tar.gz.",
             "type" : "string"
           },
+          "packagePurposes" : {
```
I'm concerned that making PackagePurpose an array will impact fidelity when translating to other formats, plus it raises ambiguity as to the "primary purpose" of a package.

goneall replied:
Since this would change the main body of the spec, I opened issue #720 to track this

goneall commented:
Thanks @zvr for the review - I updated the schemas to resolve most of the identified issues.  Comments for those issues I didn't resolve.

zvr commented:
@goneall 

Any response to my comment above:

> ` licenseId`:
> According to Annex D (SPDX license expressions), we actually allow ["DocumentRef-"(idstring)":"]"LicenseRef-"(idstring). The description should be amended to reflect this.

goneall commented:
> licenseId:
> According to Annex D (SPDX license expressions), we actually allow ["DocumentRef-"(idstring)":"]"LicenseRef-"(idstring). The description should be amended to reflect this.

@zvr - oops - missed that one.

I could not find any suitable definition in the spec that explains the semantics for the `["DocumentRef-"(idstring)":"]"LicenseRef-"(idstring)` to copy into the schema.

In thinking about this, it is a bit more complicated how to handle it in the various schemas.

Currently, the LicenseID is only used by itself as a field in the `SimpleLicenseInfo` class (parent class to the listed and extracted license classes).

In non-RDF serializations, the license expression is used to reference other licenses.  

In RDF, the URI is used to reference.

I'm thinking we only document the `ExternalDocumentRef` in the non-RDF schemas to document the license expression syntax.

The tool that generates the non-RDF schemas currently prepends the following text to all fields that use license expressions:

`License expression for [field]`.

We could expand this to something like `'License expression for [field].  See SPDX Appendix D for the license expression syntax`.

Let me know what you think.

zvr commented:
@goneall I only mentioned it because the comment of the `licenseId` field currently mentions:

> A human readable short form license identifier for a license. The license ID is either on the standard license list or the form "LicenseRef-"[idString] where [idString] is a unique string containing letters, numbers, ".", "-" or "+".

goneall commented:
@zvr How about adding the sentence:

"When used within a license expression, the license ID can optionally include a reference to an external document in the form `["DocumentRef-"(idstring)":"]"LicenseRef-"(idstring)`."

goneall commented:
@zvr I will update the license ID text to the following:

```
A human readable short form license identifier for a license. The license ID is either on the standard license list or the form "LicenseRef-[idString]" where [idString] is a unique string containing letters, numbers, ".", "-" or "+".  When used within a license expression, the license ID can optionally include a reference to an external document in the form "DocumentRef-[docrefIdString]:LicenseRef-[idString]" where docRefIdString is an ID for an external document reference.
```

Let me know if you see any issues with the above text.

goneall commented:
I also added the following to the JSON schema license expression field descriptions:

> See SPDX Annex D for the license expression syntax.

goneall commented:
> One thing -- unless I'm misunderstanding or missing the context, I don't think [idString] can include "+"? I think that the plus sign can be part of a license expression, but not part of the idString in an identifier itself.

Good catch @swinslow  - I updated the description to remove the `+`.  Updated text below.

> A human readable short form license identifier for a license. The license ID is either on the standard license list or the form "LicenseRef-[idString]" where [idString] is a unique string containing letters, numbers, "." or "-".  When used within a license expression, the license ID can optionally include a reference to an external document in the form "DocumentRef-[docrefIdString]:LicenseRef-[idString]" where docRefIdString is an ID for an external document reference.

goneall commented:
I believe all review comments have been taken care of - merging this in.
